### PR TITLE
Proposals

### LINKED ISSUE #300: Improving Voting data representation
### Problem

Currently proposals aren't very future proof in terms of different types of voting.

```js
      HashableGroupProposal (state, {data, hash}) {
        // TODO: this should be data instead of ...data to avoid conflict with neighboring properties
        // TODO: convert to votes instead of for/against for future-proofing
        state.proposals[hash] = {...data, for: [data.initiator], against: []}
      },
```

### Solution

Address those TODOs. Chat with @taoeffect before picking this up plz!

### LINKED ISSUE #343: Build out and polish generic voting / proposals system
### Problem

Need to make a generic system for submitting proposals to the group to be voted on, for various actions like adding/removing members, changing the mincome, etc.

A lot of this has been implemented but much of it isn't polished and much of it isn't hooked up to the UI.

Also keep in mind issue #192 is related to this!

### Solution

While keeping in mind related issues like #192, #300, and #312, fix the problem above.

EDIT: make sure to [check Voting.vue for improper `v-html` usage](https://github.com/okTurtles/group-income-simple/pull/493#discussion_r234628255)

### LINKED ISSUE #412: User who proposes a vote to invite a member shouldn't be asked to vote
### Problem

screen shot 2018-05-19 at 6 29 55 pm

As shown above, if bob proposes to invite someone, he shouldn't be asked to vote on his proposal again.

### Solution

Just show an "ongoing proposal" widget as @sandrina-p mocked up that lets them withdraw their vote.

### PR DESCRIPTION
**Pull request by taoeffect**
- Replaces PR #425
- Closes #343
- Implements voting rule for #596
- Closes  #412
- Closes #300
- Bump several dependencies
- Disable a few frontend tests and opened #613 to rewrite and re-enable them
- Fixed travis build failing
- Greatly improved `i18n` behavior when dealing with inline html (see #607), now you can do stuff like this:

    ```pug
    i18n(:args='{ baz }' html='foo <strong>bar</strong>: {baz}')
    ```

- Partial fix/workaround for #618
- Foundation and partial implementation for #609
- [Boyscout rule](https://github.com/okTurtles/group-income-simple/issues/383#issuecomment-383381863) for a bunch of various things

### CODE REVIEW

File: frontend/controller/router.js
Comment (sandrina-p) [inline frontend/controller/router.js:259]: If we will continue to have a mailGuard in the future, maybe it would be better to edit the guard itself  (returning always true) than removing it from where it's needed, so it's easier to add it back again.

File: frontend/model/contracts/Contract.js
Comment (sandrina-p) [inline frontend/model/contracts/Contract.js:14]: When calling a function with multiple arguments it gets hard to memorize what each one means. 
 (and even hard to maintain/evolve). Have you thought of passing an object instead? That way we don't even need to explicit pass null/undefined in the missing parameters. in this case, would be something like:

```js
return GIMessage.create({
  actionType:`${contract.name}/process`,
  actionData: data,
  metaData: metadata
})
```

File: frontend/model/contracts/group.js
Comment (sandrina-p) [inline frontend/model/contracts/group.js:198]: is this the invite code? why calling it secret? `inviteCode` seems more intuitive.

File: frontend/model/contracts/voting/proposals.js
Comment (sandrina-p) [inline frontend/model/contracts/voting/proposals.js:73]: Why passing `meta`? It seems is not used when received at `group.js`?

File: frontend/utils/time.js
Comment (sandrina-p) [inline frontend/utils/time.js:1]: Why explicitly saying strict? By default, I believe that our configs (and linters) are already in strict mode.

File: frontend/views/components/i18n.vue
Comment (sandrina-p) [inline frontend/views/components/i18n.vue:22]: It would be nice to leave documented in this file how i18n can be used (as you explained in Slack) so people can remember it easily in the future.

File: frontend/views/containers/settings/UserProfile.vue
Comment (sandrina-p) [inline frontend/views/containers/settings/UserProfile.vue]: simplified:

```
const attrCopy = this.$store.getters.currentUserIdentityContract.attributes || {};
```

File: frontend/views/pages/Contributions.vue
Comment (sandrina-p) [inline frontend/views/pages/Contributions.vue:14]: what do you mean by `XSS`? probably isn't safe because it's a text written by the user (in an input), so we could write there anything to be processed by `v-html`.

File: frontend/views/containers/settings/UserProfile.vue
Comment (sandrina-p) [inline frontend/views/containers/settings/UserProfile.vue]: You can rename now ;)

File: test/frontend.js
Comment (sandrina-p) [inline test/frontend.js]: do we really need this timeout or it was just to debug?

**Review by sandrina-p — DISMISSED** [review]
Most of my comments are questions about the details about sbp/contracts.

I believe I'll only understand this totally (and provide useful feedback related to this architecture) when I start to actually work more with it. I hope that @pieer can give valuable feedback here too.

File: frontend/views/components/GroupSettings.vue
Comment (sandrina-p) [inline frontend/views/components/GroupSettings.vue:40]: I'm okay with it, do it!

File: frontend/views/containers/modals/DesignSystem.vue
Comment (sandrina-p) [inline frontend/views/containers/modals/DesignSystem.vue]: Really? I think absolutely the opposite. I see this file as an entry point for any frontend developer to this project to get quickly familiar with our Design System and CSS framework.

File: frontend/views/containers/modals/DesignSystem.vue
Comment (sandrina-p) [inline frontend/views/containers/modals/DesignSystem.vue]: I'm fine with it, don't forget to remove these comments before merging the PR.

File: frontend/views/containers/modals/SignUp.vue
Comment (sandrina-p) [inline frontend/views/containers/modals/SignUp.vue:150]: Could you explain why are we vuex (this.$store) is being migrated to sbp (and calling the selector vuex)?

File: test/frontend.js
Comment (taoeffect) [reply test/frontend.js]: I think this was just to debug Travis, I can revert it back to 10000. 👍

File: frontend/model/contracts/Contract.js
Comment (taoeffect) [reply frontend/model/contracts/Contract.js:14]: Yes, this is a good idea and something I've given thought to myself. And since you brought it up it gives me more reason to do it. This is especially relevant for #603, which will add new options to the `create` function.

File: frontend/model/contracts/group.js
Comment (taoeffect) [reply frontend/model/contracts/group.js:198]: Because it will literally be a temporary secret that gives permission to outsiders to write messages of a restricted type to the contract.

File: frontend/model/contracts/voting/proposals.js
Comment (taoeffect) [reply frontend/model/contracts/voting/proposals.js:73]: ~~It is used by various UI widgets, for example have a look at `ProposalBox.vue`.~~

File: frontend/views/containers/modals/SignUp.vue
Comment (pieer) [reply frontend/views/containers/modals/SignUp.vue]: That's a long debate. From my understanding at the beginning

File: frontend/model/contracts/voting/proposals.js
Comment (taoeffect) [reply frontend/model/contracts/voting/proposals.js:73]: Erm, I was mistaken and you are right. It's not currently used by `'gi.contracts/group/invite'`, however all `group` messages are supposed to have metadata associated with them, so I pass along the metadata just in case it ever is used.

File: frontend/utils/time.js
Comment (taoeffect) [reply frontend/utils/time.js:1]: I believe you have to have this as a line at the top of each `.js` file... sometimes even at the top of certain functions. At least JavaScript works like that to my understanding.

File: frontend/views/components/i18n.vue
Comment (taoeffect) [reply frontend/views/components/i18n.vue:22]: Good idea, but can we leave this for #607?

File: frontend/views/containers/settings/UserProfile.vue
Comment (taoeffect) [reply frontend/views/containers/settings/UserProfile.vue]: It's not a copy if we do it that way. We don't want these values to be directly modifiable, otherwise they can get out of sync from the rest of the users.

File: frontend/views/pages/Contributions.vue
Comment (taoeffect) [reply frontend/views/pages/Contributions.vue:14]: What I mean is if this string contains any input that was generated by the user, we cannot use `v-html`, and I didn't take the time to verify whether that was the case.

File: frontend/views/containers/modals/DesignSystem.vue
Comment (taoeffect) [reply frontend/views/containers/modals/DesignSystem.vue]: OK, I just have a feeling it's going to diverge from actual usage and that might be confusing. I was thinking that maybe the actual usage (example modals throughout the project) might serve as better examples...

File: frontend/views/containers/modals/DesignSystem.vue
Comment (taoeffect) [reply frontend/views/containers/modals/DesignSystem.vue]: Why remove the comments? They're important, they're for anyone who wants to use this file as an example, to remind them to change these lines...

File: frontend/views/containers/modals/SignUp.vue
Comment (pieer) [reply frontend/views/containers/modals/SignUp.vue:150]: That's a long debate. From my understanding, sbp should be used for every action that should be saved locally and/or synchronized with the server and users. The vuex store is used for  action with no recorded state, for example open the menu. Maybe greg want also to replace completely the vuex store with sbp.

Comment (taoeffect) [reply frontend/views/containers/modals/SignUp.vue:150]: Well:

1. We might get rid of vuex completely in the future (and replace with a 100% sbp-based system), because we might switch to a different framework, see #587
2. Using `sbp` to call out to store modifications is the "right way" to do things, not because "I said so", but because SBP makes the code cleaner, improves debugging and logging, etc.
3. By using SBP it makes it possible to more easily write tests in certain situations, since selectors can be swapped out. For example, this is how we make `handleEvent` work in `test/backend.js` (by simulating it with a simplified version)

File: frontend/controller/router.js
Comment (taoeffect) [reply frontend/controller/router.js:259]: Good idea...

**Comment by taoeffect** [comment]
Note to self:

- rename one of the `UserProfile.vue` files (and component names) to `UserGroupProfile`
- change `Object.assign({}, this.$store.getters.currentUserIdentityContract.attributes || {})` to use `cloneDeep`, and also check why both files are editing the same values...
- Make both files edit the values in the same way

File: frontend/views/containers/settings/UserProfile.vue
Comment (taoeffect) [reply frontend/views/containers/settings/UserProfile.vue]: Thanks! Made a [note to self](https://github.com/okTurtles/group-income-simple/pull/611#issuecomment-521395410) in the PR and will have a closer look at the difference between the two `UserProfile.vue` files before merging.

File: frontend/views/containers/modals/SignUp.vue
Comment (taoeffect) [reply frontend/views/containers/modals/SignUp.vue:150]: Also, (4) SBP is really versatile, and works in many situations to do many different things, so it is already used all over the codebase. For this reason, it becomes simpler when writing code to just think in terms of SBP as much as possible. There are fewer concepts to deal with. I.e. it is simpler to think in terms of SBP, than to think in terms of SBP, and Vuex, and [this, that and the other]

Comment (pieer) [reply frontend/views/containers/modals/SignUp.vue:150]: Now, calling the selector 'state/vuex/dispatch' is a good reminder that when we will remove vuex completely we need to check this event.

File: frontend/utils/time.js
Comment (pieer) [reply frontend/utils/time.js:1]: Not a problem for me. It can be a useful information for some editor.

File: frontend/views/components/GroupSettings.vue
Comment (pieer) [reply frontend/views/components/GroupSettings.vue:40]: This is a useful temporary file until we move all the functionalities present in that component somewhere (only partially done atm).

File: frontend/model/contracts/voting/proposals.js
Comment (sandrina-p) [reply frontend/model/contracts/voting/proposals.js:73]: I see... but if it's not needed now, why passing "noise". We should keep it clear/simple/pass only the essential. Anyway, could you explain to me a little more about what this `metadata` could contain?

File: frontend/utils/time.js
Comment (sandrina-p) [reply frontend/utils/time.js:1]: I think it's just "noise" because babel already does that by default for us.

@pieer what editor are you talking about? As far as I know for VScode, Atom and Sublime it does nothing 🤔

File: frontend/model/contracts/voting/proposals.js
Comment (taoeffect) [reply frontend/model/contracts/voting/proposals.js:73]: It's not noise at all, it's part of what is essential. Since the group contract defines metadata, it means every message must have metadata attached to it. If I were to delete this line, it would throw an exception when the message is validated. Look at the group contract definition in `group.js` to see what `meta` must contain.

File: frontend/views/containers/modals/DesignSystem.vue
Comment (sandrina-p) [reply frontend/views/containers/modals/DesignSystem.vue]: Personally having "demos" is always a good entry point to any feature. That's the purpose of DS page, to be a "demo" of every component we have (not only modals, but also forms, messages, etc...)

File: frontend/utils/time.js
Comment (taoeffect) [reply frontend/utils/time.js:1]: We want to follow the JS language specs explicitly if we can, and put this at the top of every file for performance and safety reasons. Plus, since we might one day remove `babel` as a dependency, that's all the more reason to do it.

File: frontend/views/containers/modals/SignUp.vue
Comment (sandrina-p) [reply frontend/views/containers/modals/SignUp.vue:150]: Thank you guys for the clarification. This intel is too good to be lost in a PR. It would be nice to add this info as documentation. Maybe at `model/state.js` it would be a good place.

File: frontend/model/contracts/voting/proposals.js
Comment (sandrina-p) [reply frontend/model/contracts/voting/proposals.js:73]: It needs `metadata` to create a group, but it doesn't need to invite someone... But I understand your point.

Resolving this.

File: frontend/views/containers/settings/UserProfile.vue
Comment (sandrina-p) [reply frontend/views/containers/settings/UserProfile.vue]: Oh, right, didn't thought about the clone part. So, you can simply use destructuring:

```
const attrCopy = { ...this.$store.getters.currentUserIdentityContract.attributes };
```

File: frontend/utils/time.js
Comment (pieer) [reply frontend/utils/time.js:1]: If you want to debug a single file of code online or in the browser for example. But really it doesn't matter. If you want to remove, let's remove it.

File: frontend/model/contracts/Contract.js
Comment (taoeffect) [reply frontend/model/contracts/Contract.js:14]: Resolved via an edit to #603

File: frontend/controller/router.js
Comment (taoeffect) [reply frontend/controller/router.js:259]: Will resolve this by simply commenting out that line and the guard both (since it's no longer used, and unused vars is a linter error)

File: frontend/views/components/GroupSettings.vue
Comment (taoeffect) [reply frontend/views/components/GroupSettings.vue:40]: I'll leave this up to you @sandrina-p and @pieer to figure out. :)

File: frontend/views/containers/modals/SignUp.vue
Comment (taoeffect) [reply frontend/views/containers/modals/SignUp.vue:150]: I'm not sure that `model/state.js` is the right place to put all this documentation... however maybe it could go somewhere in the `docs/` folder?

File: frontend/model/state.js
Comment (pieer) [inline frontend/model/state.js]: Why did you removed the font size?

**Review by pieer — CHANGES_REQUESTED** [review]

File: frontend/model/state.js
Comment (taoeffect) [reply frontend/model/state.js]: It wasn't being used anywhere, and 1 is a weird default font size.

Comment (pieer) [reply frontend/model/state.js]: It's useful for the fontSelector, please keep the variable and the functions associated to it.
1 is not weird. it's at the opposite a good default value.

Comment (taoeffect) [reply frontend/model/state.js]: OK, I can add it back, but there needs to be a comment explaining why it's 1. What unit is it?

Comment (pieer) [reply frontend/model/state.js]: it's not a unit it's a scale

Comment (pieer) [reply frontend/model/state.js]: I could have put small medium and big, but then it would have limit the user to only 3 values

Comment (sandrina-p) [reply frontend/model/state.js]: @greg The default is 1 for now, but it's used so the user can customize the web font-size in the user preferences in the future

**Review by pieer — APPROVED** [review]

**Pull request merged**
